src/main/java/org/jenkinsci/plugins/workflow/actions/ArgumentsAction.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.actions;
+
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graph.StepNode;
+import org.jenkinsci.plugins.workflow.steps.Step;
+import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Stores some or all of the arguments used to create and configure the {@link Step} executed by a {@link FlowNode}.
+ * This allows you to inspect information supplied in the pipeline script and otherwise discarded at runtime.
+ * Supplied argument values can be hidden and replaced with a {@link NotStoredReason} for security or performance.
+ */
+public abstract class ArgumentsAction implements PersistentAction {
+
+    /** Used as a placeholder for arguments not stored for various reasons */
+    public enum NotStoredReason {
+        /** Denotes an unsafe value that cannot be stored/displayed due to sensitive info */
+        MASKED_VALUE,
```

Comment:
Note that this API design basically precludes a finer-grained behavior if that is desirable in the future. For example, if we decide that it is valuable to retain everything except the actual password in a step like

```groovy
withCredentials(…) {
  sh "mvn -DscmPass=$SCMPASS -B release:prepare release:perform"
}
```

(not itself a great example since use of `'` over `"` is more advisable here), the implementation would have no way of representing that: it could return a `String` that is partially masked out, but callers could not know this status.

Similarly, if we decided we wanted

```groovy
sh """
make world
if [ -f errors ]
then
  echo long and complicated script follows, far beyond maximum parameter length
fi
"""
```

to be translated to something like

```java
"make world\n…"
```

or whatever, rather than returning `OVERSIZE_VALUE`, an implementation could not do that without also informing clients that this had been done and that they should not expect the argument to be complete.

Whether this is really a problem, I am not sure, but wanted to flag the limitation.